### PR TITLE
configure_file: make input arg optional if command is used

### LINKED ISSUE #1476: configure_file: make input kwarg optional if command is specified?
I'm using configure_file() in the Gtk+ meson build to generate gresource.xml files based on a bunch of input filenames. This is done by calling a python script and passing the output filename and the input files to it.

I wonder if configure_file() really needs to insist on an input kwarg for this case, where `command:` is used?

### PR DESCRIPTION
**Pull request by tp-m**
Fixes #1476

### CODE REVIEW

File: mesonbuild/interpreter.py [hunk 1]
```diff
@@ -2277,7 +2278,11 @@ def func_configure_file(self, node, args, kwargs):
             if not isinstance(inputfile, (str, mesonlib.File)):
                 raise InterpreterException('Input must be a string or a file')
             ifile_abs = os.path.join(self.environment.source_dir, self.subdir, inputfile)
-        elif 'command' in kwargs:
+        elif 'command' in kwargs and not '@INPUT@' in kwargs['command']:
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E713] test for membership should be 'not in'

File: mesonbuild/interpreter.py [hunk 2]
```diff
@@ -2277,7 +2278,11 @@ def func_configure_file(self, node, args, kwargs):
...
+          pass
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 3]
```diff
@@ -2277,7 +2278,11 @@ def func_configure_file(self, node, args, kwargs):
...
+        elif 'configuration' in kwargs:
+          pass
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 4]
```diff
@@ -2307,7 +2312,10 @@ def func_configure_file(self, node, args, kwargs):
             # We use absolute paths for input and output here because the cwd
             # that the command is run from is 'unspecified', so it could change.
             # Currently it's builddir/subdir for in_builddir else srcdir/subdir.
-            values = mesonlib.get_filenames_templates_dict([ifile_abs], [ofile_abs])
+            if ifile_abs:
+              values = mesonlib.get_filenames_templates_dict([ifile_abs], [ofile_abs])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py [hunk 5]
```diff
@@ -2307,7 +2312,10 @@ def func_configure_file(self, node, args, kwargs):
...
+            else:
+              values = mesonlib.get_filenames_templates_dict(None, [ofile_abs])
```
Comment (ignatenkobrain) [inline mesonbuild/interpreter.py]: __[flake8]__
* [E111] indentation is not a multiple of four

File: mesonbuild/interpreter.py (hunk 1 above)
Comment (nirbheek) [inline mesonbuild/interpreter.py]: Why not make this:

```python
elif 'command' in kwargs and '@INPUT@' in kwargs['command']:
    raise InterpreterException(...)
```

Comment (nirbheek) [inline mesonbuild/interpreter.py]: This will also ensure that people won't pass `@INPUT@` without `input:`.

File: mesonbuild/interpreter.py [hunk 6]
```diff
@@ -2307,7 +2312,10 @@ def func_configure_file(self, node, args, kwargs):
             # We use absolute paths for input and output here because the cwd
             # that the command is run from is 'unspecified', so it could change.
             # Currently it's builddir/subdir for in_builddir else srcdir/subdir.
-            values = mesonlib.get_filenames_templates_dict([ifile_abs], [ofile_abs])
+            if ifile_abs:
+                values = mesonlib.get_filenames_templates_dict([ifile_abs], [ofile_abs])
+            else:
+                values = mesonlib.get_filenames_templates_dict(None, [ofile_abs])
```
Comment (nirbheek) [inline mesonbuild/interpreter.py]: I was actually thinking that it makes sense to allow people to pass any number of files as input, so `ifile_abs` would then come from `meson.stringlistify(kwargs.get('input', []))`, would either be N files list or an empty list, and we won't need an if condition here.

Comment (tp-m) [inline mesonbuild/interpreter.py]: Maybe, but I think this should be a separate commit/PR then. The current code clearly allows only a single input file intentionally.

File: test cases/common/16 configure file/meson.build [hunk 7]
```diff
@@ -29,13 +29,23 @@ ifile = '@0@/dummy.dat'.format(meson.current_source_dir())
...
   install_dir : 'share/appdir')
 run_command(check_file, join_paths(meson.current_build_dir(), 'config2.h'))
 
+# Same again, but an input file should not be required in this
+# case where we use a command to generate the output file.
+ofile2b = '@0@/config2b.h'.format(meson.current_build_dir())
+configure_file(
+  output : 'config2b.h',
+  command : [genprog, scriptfile, ifile, ofile2b],
+  install_dir : 'share/appdir')
```
Comment (jpakkane) [inline test cases/common/16 configure file/meson.build]: It is confusing that the comment says that an input file is not needed but still it is given. There should be a different script that actually does not take inputs or the generator script should be changed so that it can take either 2 or 3 arguments.

**Comment by tp-m** [comment]
Updated.
